### PR TITLE
Feature/sable cc grammar

### PR DESCRIPTION
**Pull request by KarmaKamikaze**

### CODE REVIEW

**Comment by KarmaKamikaze** [comment]
[Convert CFG draft to SableCC grammar](https://trello.com/c/N2CWFQcs/52-convert-cfg-draft-to-sablecc-grammar)

File: STEP.grammar [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+Productions
+  // Program
+  program = setuploop P.functions?;
+  setuploop = {one} P.setup | {two} P.loop | {three} P.setup P.loop;
+  setup = T.setup stmt* endsetup nl;
+  loop = T.loop stmt* endloop nl;
+  
+  // Function declarations
+  functions = T.functions funcdcl* endfunctions nl;
+  funcdcl = {one} type brackets? T.function id params stmt* retstmt endfunction nl | {two} blank T.function id params stmt* endfunction nl;
```
Comment (Sarmisuper) [inline STEP.grammar]: Enable return of more than one dimensional arrays

Comment (mirakst) [inline STEP.grammar]: single dimensional array only
proposed code allows multiple dimensions
```suggestion
  funcdcl = {one} type brackets* T.function id params stmt* retstmt endfunction nl | {two} blank T.function id params stmt* endfunction nl;
```

File: STEP.grammar [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  forstmt_iterator_options = {one} numdcl | {two} assstmt | {three} id arrindex?;
+  option_increase_decrease = {one} increaseby | {two} decreaseby;
+  switchstmt = switch lparen id arrindex? rparen whencase* otherwisecase? endswitch;
+  whencase = when numliteral do stmt* fallthrough?;
+  otherwisecase = otherwisedo stmt*;
+  assstmt = id arrindex? assign expr;
+  funccall = id lparen params_options? rparen;
+  params_options = expr multi_expr*;
+  multi_expr = comma expr;
+  retstmt = {one} return cond;
```
Comment (mirakst) [inline STEP.grammar]: allow return in blank functions
```suggestion
  retstmt = {one} return cond | {two} return;
```

File: STEP.grammar [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  arrindex = lbrack expr rbrack;
+  skipstmt = nl;
+  
+  // Arithmetic expressions
+  expr = {one} value | {two} expr aop value;
+  aop = {one} plus | {two} minus | {three} mult | {four} divide | {five} pow;
+  value = {one} P.constant | {two} id | {three} lparen expr rparen;
```
Comment (mirakst) [inline STEP.grammar]: allow array access in expressions
```suggestion
  value = {one} P.constant | {two} id arrindex? | {three} lparen expr rparen;
```

File: STEP.grammar [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  constant = {one} numliteral | {two} strliteral | {three} boolliteral;
```
Comment (Sarmisuper) [inline STEP.grammar]: Missing operator precendence

File: STEP.grammar [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  bop = {one} and | {two} or;
+  
+  // Variable declarations
+  vardcl = T.constant? var_options nl;
+  var_options = {one} numdcl | {two} stringdcl | {three} booldcl | {four} arrdcl;
+  numdcl = number id assign expr nl;
+  stringdcl = string id assign expr nl;
+  booldcl = boolean id assign cond nl;
+  arrdcl = type [fst]:arrsizedcl [multi]:arrsizedcl* id assign arr_id_or_expr;
+  arr_id_or_expr = {one} id | {two} lbrack params_options? rbrack;
```
Comment (Sarmisuper) [inline STEP.grammar]: does not allow for multidimensional arrays

**Review by Sarmisuper — APPROVED** [review]

File: STEP.grammar (hunk 5 above)
Comment (mirakst) [inline STEP.grammar]: allow multidimensional array literal
```suggestion
 arr_id_or_expr = {one} id | {two} lbrack arr_lit* rbrack;
 arr_lit = {one} params_options? | {two}  lbrack params_options? rbrack | {three} lbrack params_options? rbrack comma 
 arr_lit;
```

**Review by mirakst — CHANGES_REQUESTED** [review]
very nice and cool and amazing good job
pls adds suggestions

File: STEP.grammar (hunk 4 above)
Comment (KarmaKamikaze) [reply STEP.grammar]: We will fix it after testing

**Review by mirakst — APPROVED** [review]

**Comment by KarmaKamikaze** [comment]
[Develop a lexer/scanner](https://trello.com/c/iUU1mWJF/8-develop-a-lexer-scanner)

**Pull request merged**